[CUB-2066] extract eslint and prettier config into a shared repo

ignatiusreza opened the pull request:
this is based on the config used by Laboperator, adjusted to fit with the one used by Labfolder (airbnb-config-base)..

Some eslint stylistic configuration are not included, since I think it will be covered by prettier..

For Laboperator team, some rules used by Labfolder team are now included by default.. I judged that these rules shouldn't affect us much and is good to have anyway.. diff below for the differences:

```diff
diff --git a/config/eslint/base.yml b/config/eslint/base.yml
index 07ad17154..bfaeb4ce1 100644
--- a/config/eslint/base.yml
+++ b/config/eslint/base.yml
@@ -5,46 +5,137 @@ parser: babel-eslint
 rules:
   # https://eslint.org/docs/rules/
   # Possible Errors
+  getter-return:
+    - error
+    - allowImplicit: true
+  no-await-in-loop: error
+  no-cond-assign: ['error', 'always']
   no-console: warn
+  no-template-curly-in-string: error
+  require-atomic-updates: off
   # Best Practices
+  array-callback-return:
+    - error
+    - allowImplicit: true
+  block-scoped-var: error
+  consistent-return: error
   complexity: ['warn', 10]
   curly: ['error', 'multi-line', 'consistent']
+  default-case: error
+  default-case-last: error
+  default-param-last: error
   dot-location: ['error', 'property']
   dot-notation: error
   eqeqeq: ['error', 'smart']
   no-alert: error
+  no-caller: error
+  no-constructor-return: error
+  no-empty-function:
+    - error
+    - allow: ['arrowFunctions', 'functions', 'methods']
+  no-eval: error
   no-extend-native: error
+  no-extra-bind: error
+  no-extra-label: error
+  no-floating-decimal: error
+  no-global-assign: error
   no-implicit-globals: error
+  no-iterator: error
+  no-labels:
+    - error
+    - allowLoop: false
+      allowSwitch: false
   no-lone-blocks: error
-  no-native-reassign: error
+  no-loop-func: error
+  no-new: error
+  no-new-func: error
+  no-new-wrappers: error
+  no-octal-escape: error
+  no-param-reassign:
+    - error
+    - props: true
+      ignorePropertyModificationsFor:
+        - 'acc' # for reduce accumulators
+        - 'accumulator' # for reduce accumulators
+        - 'e' # for e.returnvalue
+        - 'ctx' # for Koa routing
+        - 'req' # for Express requests
+        - 'request' # for Express requests
+        - 'res' # for Express responses
+        - 'response' # for Express responses
+        - '$scope' # for Angular 1 scopes
+        - 'staticContext' # for ReactRouter context
+  no-proto: error
+  no-return-assign: ['error', 'always']
+  no-return-await: error
+  no-script-url: error
+  no-self-assign: error
+  no-self-compare: error
+  no-sequences: error
+  no-throw-literal: error
+  no-unused-expressions: error
   no-useless-concat: error
+  no-useless-return: error
+  no-void: error
+  radix: error
   vars-on-top: error
+  wrap-iife:
+    - error
+    - outside
+    - functionPrototypeMethods: false
   yoda: error
+  # Strict
+  strict: ['error', 'never']
   # Variables
   no-label-var: error
   no-restricted-globals: ['error', 'event']
+  no-shadow: error
+  no-undef-init: error
   no-unused-vars:
     - error
     - args: after-used
-      vars: local
       argsIgnorePattern: ^_
+      ignoreRestSiblings: true
       varsIgnorePattern: ^_
+  no-use-before-define:
+    - error
+    - functions: false
   # Stylistic Issues
   camelcase:
     - warn
     - properties: always
-  consistent-this: ['error', 'that']
+  func-names: warn
   func-style:
     - error
     - declaration
     - allowArrowFunctions: true
+  implicit-arrow-linebreak: ['error', 'beside']
   max-depth: ['warn', 4]
   max-nested-callbacks: ['warn', 3]
   max-params: ['warn', 4]
   max-statements: ['warn', 10]
   new-cap: error
+  new-parens: error
+  no-array-constructor: error
+  no-bitwise: error
+  no-continue: error
   no-lonely-if: error
+  no-mixed-operators:
+    - error
+    - allowSamePrecedence: false
+      groups:
+        - ['%', '**']
+        - ['%', '+']
+        - ['%', '-']
+        - ['%', '*']
+        - ['%', '/']
+        - ['/', '*']
+        - ['&', '|', '<<', '>>', '>>>']
+        - ['==', '!=', '===', '!==']
+        - ['&&', '||']
+  no-multi-assign: error
   no-nested-ternary: error
+  no-new-object: error
+  no-plusplus: error
   no-unneeded-ternary: error
   operator-assignment: ['error', 'always']
   padding-line-between-statements:
@@ -57,6 +148,7 @@ rules:
         'next': ['const', 'let', 'var'],
       },
     ]
+  prefer-object-spread: error
   quotes:
     - error
     - single
@@ -71,13 +163,25 @@ rules:
         - '='
       markers:
         - '='
+        - '!'
+        - '/'
+  unicode-bom: ['error', 'never']
   # ECMAScript 6
-  no-var: warn
+  arrow-body-style: error
+  arrow-parens: error
+  no-confusing-arrow: error
+  no-useless-computed-key: error
+  no-useless-constructor: error
+  no-useless-rename: error
+  no-var: error
+  object-shorthand: error
+  prefer-arrow-callback: error
   prefer-const:
     - error
     - destructuring: all
+  prefer-destructuring: error
+  prefer-numeric-literals: error
+  prefer-rest-params: error
   prefer-spread: error
-  prefer-template: warn
-  no-use-before-define:
-    - error
-    - functions: false
+  prefer-template: error
+  symbol-description: error
```

For Labfolder team, there're several rules in the `airbnb-config-base` which I feel are too restrictive.. so, I choose to exclude them for now.. If anyone feels strongly that it should be part of the shared config, please open separate PR for each rules and we can discuss each one individually.. Several rules which are used by Laboperator but are not part of the `airbnb-config-base` is also now included by default.. you can add the following configuration to get the same configuration as what you're using now:

```js
const confusingBrowserGlobals = require('confusing-browser-globals');

{
  'complexity': ['off', 11],
  'no-else-return': ['error', { allowElseIf: false }],
  'no-implicit-globals': 'off',
  'no-restricted-properties': ['error', {
    object: 'arguments',
    property: 'callee',
    message: 'arguments.callee is deprecated',
  }, {
    object: 'global',
    property: 'isFinite',
    message: 'Please use Number.isFinite instead',
  }, {
    object: 'self',
    property: 'isFinite',
    message: 'Please use Number.isFinite instead',
  }, {
    object: 'window',
    property: 'isFinite',
    message: 'Please use Number.isFinite instead',
  }, {
    object: 'global',
    property: 'isNaN',
    message: 'Please use Number.isNaN instead',
  }, {
    object: 'self',
    property: 'isNaN',
    message: 'Please use Number.isNaN instead',
  }, {
    object: 'window',
    property: 'isNaN',
    message: 'Please use Number.isNaN instead',
  }, {
    property: '__defineGetter__',
    message: 'Please use Object.defineProperty instead.',
  }, {
    property: '__defineSetter__',
    message: 'Please use Object.defineProperty instead.',
  }, {
    object: 'Math',
    property: 'pow',
    message: 'Use the exponentiation operator (**) instead.',
  }],
  'prefer-promise-reject-errors': ['error', { allowEmptyReject: true }],
  'no-restricted-globals': ['error', 'isFinite', 'isNaN'].concat(confusingBrowserGlobals),
  'max-depth': ['off', 4],
  'max-len': ['error', 100, 2, {
    ignoreUrls: true,
    ignoreComments: false,
    ignoreRegExpLiterals: true,
    ignoreStrings: true,
    ignoreTemplateLiterals: true,
  }],
  'max-nested-callbacks': 'off',
  'max-params': ['off', 3],
  'max-statements': ['off', 10],
  'newline-per-chained-call': ['error', { ignoreChainWithDepth: 4 }],
  'no-restricted-syntax': [
    'error',
    {
      selector: 'ForInStatement',
      message: 'for..in loops iterate over the entire prototype chain, which is virtually never what you want. Use Object.{keys,values,entries}, and iterate over the resulting array.',
    },
    {
      selector: 'ForOfStatement',
      message: 'iterators/generators require regenerator-runtime, which is too heavyweight for this guide to allow them. Separately, loops should be avoided in favor of array iterations.',
    },
    {
      selector: 'LabeledStatement',
      message: 'Labels are a form of GOTO; using them makes code confusing and hard to maintain and understand.',
    },
    {
      selector: 'WithStatement',
      message: '`with` is disallowed in strict mode because it makes code impossible to predict and optimize.',
    },
  ],
  'padding-line-between-statements': 'off',
}
```

ignatiusreza commented:
Corresponding PR in laboperator-server: https://github.com/labforward/laboperator-server/pull/1661

doktordirk commented:
```js
{
      selector: 'LabeledStatement',
      message: 'Labels are a form of GOTO; using them makes code confusing and hard to maintain and understand.',
    },
```
you  use labels?

doktordirk commented:
quick look at the optionals: only one which tends to annoy me is `'consistent-return': error,`. all others would be fine with me it seems. and the consistent return would be kinda required for typescript anyways ..

ignatiusreza commented:
> you use labels?

nope, it's in the `airbnb-config-base` for some reason.. :shrug: 

> only one which tends to annoy me is `'consistent-return': error,`.

yup, I also was not sure if this should be default or not.. I guess in typescript, you can also kinda do inconsistent return by doing `(): A | B | C`? not sure..

right now, laboperator code base is somewhat littered with this issue, which hopefully will gets better once we officially adopt typescript..

ignatiusreza commented:
thinking about it again.. let's activate `consistent-return: error` by default.. it's the right things to do, and as mentioned is required by typescript anyway..

doktordirk commented:
fyi: applying it to labfolder/laboperator https://github.com/labforward/labfolder-web/pull/4540 (we had been neglecting it a bit)

doktordirk commented on eslint/cypress.yml:
```diff
@@ -0,0 +1,8 @@
+extends:
+  - "./react.yml"
```
no react for us to be shareable

ignatiusreza replied:
Nice catch! Sorry I missed this.. removed from default config.. we'll extends react manually in laboperator-server

hiaux0 commented:
>  If anyone feels strongly that it should be part of the shared config, please open separate PR

Sounds like the only sane way to handle this. :+1:

ignatiusreza commented:
@doktordirk & @hiaux0 okay to move forward with this?

doktordirk commented:
there's an issue with eslint. several multiline binary expresions get mangled 
before:
```js
export function unpatchCommands(originalCommands) {
  Object.keys(GC.Spread.Sheets.Commands)
    .filter(entryName => typeof GC.Spread.Sheets.Commands[entryName].execute === 'function'
      && !commandPatchBlackList.includes(entryName))
    .forEach((commandName) => {
      GC.Spread.Sheets.Commands[commandName].execute = originalCommands[commandName];
    });
```

after (`&&` removed)
```js
  Object.keys(GC.Spread.Sheets.Commands)
    .filter((entryName) => typeof GC.Spread.Sheets.Commands[entryName].execute === 'function'
        !commandPatchBlackList.includes(entryName)
    )
    .forEach((commandName) => {
      GC.Spread.Sheets.Commands[commandName].execute =
        originalCommands[commandName];
    });
```

there was another occasion where `.filter` got just removed!? (dunno if i find it again).

so before we can apply it we'd better find out what's the issue. a workaround for our local eslint seems to be to add `'implicit-arrow-linebreak': 'off',`

ignatiusreza commented:
oh, interesting.. `implicit-arrow-linebreak` rules comes from the `airbnb-base` lint.. so I thought it should already be part of your current config.. but, actually the `airbnb-base` config used by labfolder is quite outdated.. I'll double check the rules..

ignatiusreza commented:
@doktordirk I removed `implicit-arrow-linebreak` from the default rules, also adjusted `spaced-comment` to allow `////////` and `/***********/`.. 

> there was another occasion where .filter got just removed!? (dunno if i find it again).

I checked your PR in labfolder-web, but can't find this issue.. would be great to get some pointer on this..

doktordirk commented on eslint/base.yml:
```diff
@@ -0,0 +1,188 @@
+  unicode-bom: ['error', 'never']
+  # ECMAScript 6
+  arrow-body-style: error
+  arrow-parens: error
```
i narrowed the problem down to this one. with `arrow-parens: ["error", "as-needed"]` or `arrow-parens: ["error", "as-needed", { "requireForBlockBody": true }]` it works fine though

ignatiusreza replied:
interesting.. changing this to `as-needed` would result in conflicting format with prettier..

I think ultimately we can remove this rule, since it is covered by [this rule](https://prettier.io/docs/en/options.html#arrow-function-parentheses) in prettier.. but, prettier also have the same default to always include parentheses.. does it breaks if it's auto formatted by prettier?

ignatiusreza replied:
Tested locally with `arror-parens` disabled, and it seems to work, prettier format it as intended, and the resulting output is valid (`.filter` are not removed)..

so should be good to go?

doktordirk commented:
the file where eslint removes `.filter`: `labfolder-eln-webapp/cypress/support/playground/unique.js`

ignatiusreza commented:
Awesome! Thanks a lot!
